File: src/main/java/io/github/abductcows/easyargs/ArgumentParser.java
/*
   Copyright 2022 George Bouroutzoglou

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

package io.github.abductcows.easyargs;

import io.github.abductcows.easyargs.ArgumentParseException.BadArgumentUseException;
import io.github.abductcows.easyargs.ArgumentParseException.DuplicateArgumentNameException;
import io.github.abductcows.easyargs.ArgumentParseException.ParsingNotFinishedException;
import org.jetbrains.annotations.Nullable;

import java.util.*;

import static io.github.abductcows.easyargs.Utils.*;

/**
 * Parser object that parses the program arguments, looking for programmer defined ones.
 * <p>
 * Returns a queryable result that returns information about the program arguments.
 */
@CustomNonNullAPI
public final class ArgumentParser {

    private Map<String, Argument> argsLookupByName = new HashMap<>();
    private ArgumentParserResult result = new ArgumentParserResult();
    private boolean parsingFinished = false;

    @Nullable
    private RuntimeException parseException = null;

    /**
     * Argument parser constructor
     * <p>
     * Parser objects can be reused as new after a call to {@link #getParserResult}
     */
    public ArgumentParser() {
    }

    /**
     * Parses the program arguments and returns the queryable result used to handle the arguments
     * <p>
     * All exceptions are stored and thrown after parsing has finished. If caught, the remaining valid arguments
     * are available via {@link #getParserResult()}. If multiple exceptions occur, the first one is thrown, with the others
     * suppressed
     * </p>
     *
     * @param programArgs the program arguments
     * @param myArgs      programmer defined arguments
     * @return the result of argument parsing
     * @throws BadArgumentUseException        if an expected argument is supplied (same name) but does not conform to the full
     *                                        specification (e.g. --port without value). The offending argument is discarded.
     * @throws DuplicateArgumentNameException if two arguments are found with the same short or long name. The first argument will be kept.
     */
    public ArgumentParserResult parseForMyArgs(String[] programArgs, List<Argument> myArgs) {

        List<Argument> argsCopy = discardDuplicates(myArgs);
        populateLookupTable(argsCopy);

        for (int i = 0; i < programArgs.length; i++) {

            String currentArg = programArgs[i];
            if (!isValidArgumentName(currentArg)) continue;

            Argument currentArgument = argsLookupByName.get(currentArg);
            if (assertProperlyUsed(currentArgument, programArgs, i)) {
                storeArgumentInResult(currentArgument, programArgs, i);
            }

            removeFromLookupTable(currentArgument);
        }

        parsingFinished = true;
        if (parseException != null) {
            throw parseException;
        }
        return result;
    }

    private List<Argument> discardDuplicates(List<Argument> myArgs) {

        Set<String> argNamesSoFar = new LinkedHashSet<>();
        List<Argument> uniqueArgs = new ArrayList<>();

        for (Argument arg : myArgs) {

            if (argNamesSoFar.contains(arg.getShortName()) || argNamesSoFar.contains(arg.getLongName())) {
                String argumentName = returnForFirstNonEmptyName(arg, String::toString);
                storeException(new DuplicateArgumentNameException(argumentName));
            } else {
                uniqueArgs.add(arg);
                runForNonEmptyNames(arg, argNamesSoFar::add);
            }
        }

        return uniqueArgs;
    }

    /**
     * Vararg version of {@link #parseForMyArgs(String[], List)}
     *
     * @param programArgs the program arguments
     * @param myArgs      programmer defined arguments
     * @return the result of argument parsing
     */
    public ArgumentParserResult parseForMyArgs(String[] programArgs, Argument... myArgs) {
        return parseForMyArgs(programArgs, Arrays.asList(myArgs));
    }

    /**
     * Returns the result of the last argument parsing.
     * <p>
     * The result is returned only once. Subsequent calls without {@link #parseForMyArgs}
     * will return empty {@link ArgumentParserResult} objects.
     * </p>
     *
     * @return the queryable result
     * @throws ParsingNotFinishedException if called before {@link #parseForMyArgs}
     */
    public ArgumentParserResult getParserResult() {
        if (!parsingFinished) throw new ParsingNotFinishedException();
        ArgumentParserResult result = this.result;
        reset();
        return result;
    }


    private void storeArgumentInResult(Argument argument, String[] programArgs, int indexInProgramArgs) {
        if (argument.getNeedsValue()) {
            String value = programArgs[indexInProgramArgs + 1];
            result.addArgumentWithValue(argument, value);
        } else {
            result.addSimpleArgument(argument);
        }
    }

    private void populateLookupTable(List<Argument> myArgs) {

        for (Argument argument : myArgs) {
            runForNonEmptyNamesAddingHyphens(argument, argumentNameWithHyphens -> {
                argsLookupByName.put(argumentNameWithHyphens, argument);
            });
        }
    }

    private boolean isValidArgumentName(String currentArg) {
        return argsLookupByName.containsKey(currentArg);
    }

    private boolean assertProperlyUsed(Argument argument, String[] programArgs, int indexInProgramArgs) {


        if (argument.getNeedsValue()) {
            int indexOfValue = indexInProgramArgs + 1;
            if (indexOfValue >= programArgs.length ||
                    isValidArgumentName(programArgs[indexOfValue])) {
                String readableName = Utils.returnForFirstNonEmptyName(argument, String::toString);
                String nextValue = indexOfValue >= programArgs.length ? "" : programArgs[indexOfValue];
                storeException(new BadArgumentUseException(readableName + " requires a value, but \"" + nextValue +
                        "\" cannot be used as a value"));
                return false;
            }
        }

        return true;
    }

    private void storeException(RuntimeException e) {
        if (parseException == null) {
            parseException = e;
        } else {
            parseException.addSuppressed(e);
        }
    }

    private void removeFromLookupTable(Argument argument) {
        runForNonEmptyNamesAddingHyphens(argument, argsLookupByName::remove);
    }

    private void reset() {
        argsLookupByName = new HashMap<>();
        result = new ArgumentParserResult();
        parsingFinished = false;
        parseException = null;
    }
}

File: src/main/kotlin/io/github/abductcows/easyargs/parser/ArgumentParser.kt
/*
   Copyright 2022 George Bouroutzoglou

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */
@file:Suppress("unused")

package io.github.abductcows.easyargs.parser

import io.github.abductcows.easyargs.arguments.Argument

/**
 * Parses the program arguments, looking for programmer defined ones.
 *
 * Returns a queryable result that can return:
 * - whether an argument has been supplied
 * - if it is an argument-value pair, the value
 */
class ArgumentParser {

    private var argsLookupByName: MutableMap<String, Argument> = HashMap()
    private var result: ArgumentParserResult = ArgumentParserResult()
    private var parsingFinished: Boolean = false

    /**
     * Parses the program arguments String array, looking for occurrences of the arguments provided
     *
     * Any other arguments are left intact and will not throw an exception. The parser finishes parsing and then
     * throws any exceptions related to the arguments specified. If you recover from this exception, the valid arguments
     * will still be available by calling [getParseResult]
     *
     * @throws BadArgumentUseException if a String matching an [argument][Argument] violates its specification. E.g.
     * does not have a value while [Argument.getNeedsValue] is true
     */
    fun parseForMyArgs(programArgs: Array<String>, myArgs: List<Argument>): ArgumentParserResult {
        var parseException: Exception? = null

        // populate the lookup table
        populateLookupTable(myArgs)

        // parse the arguments
        for (i in programArgs.indices) {
            val currentString = programArgs[i]

            // skip non-specified arguments
            if (!isValidArgumentName(currentString)) continue

            // valid arg name, add if used properly or throw on bad use
            val currentArgument = argsLookupByName.getValue(currentString)
            if (!isProperlyUsed(currentArgument, i, programArgs)) {
                val message = "Sample message" // todo add a proper message depending on argument
                if (parseException == null) {
                    parseException = BadArgumentUseException(message)
                } else {
                    parseException.addSuppressed(BadArgumentUseException(message))
                }
                continue
            }

            // add it
            if (currentArgument.needsValue) {
                result.addArgWithValue(currentArgument, programArgs[i + 1])
            } else {
                result.addSimpleArgument(currentArgument)
            }
        }

        parsingFinished = true
        parseException?.let { throw it }
        return result
    }

    /**
     * Vararg overload of [parseForMyArgs]
     */
    fun parseForMyArgs(programArgs: Array<String>, vararg definedArgs: Argument) =
        parseForMyArgs(programArgs, definedArgs.toList())

    /**
     * Returns the results of the argument parsing. Called after [parseForMyArgs]
     * @throws ParsingNotFinishedException if called before [parseForMyArgs]
     */
    fun getParseResult(): ArgumentParserResult {
        if (!parsingFinished) throw ParsingNotFinishedException()
        return result
    }

    private fun populateLookupTable(myArgs: List<Argument>) {
        for (argument in myArgs) {
            if (argument.shortName.isNotEmpty()) {
                argsLookupByName["-${argument.shortName}"] = argument
            }
            if (argument.longName.isNotEmpty()) {
                argsLookupByName["--${argument.longName}"] = argument
            }
        }
    }

    private fun isValidArgumentName(arg: String) = argsLookupByName.contains(arg)
    private fun isProperlyUsed(argument: Argument, argumentIndex: Int, programArgs: Array<String>): Boolean {

        // check if it needs a value
        if (argument.needsValue) {
            if (argumentIndex + 1 !in programArgs.indices || // no more args
                isValidArgumentName(programArgs[argumentIndex + 1]) // or next one is another argument
            ) {
                return false
            }
        }

        return true
    }

}

